expression/builtin_string.go

```diff
@@ -426,33 +426,41 @@ type lowerFunctionClass struct {
 }
 
 func (c *lowerFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinLowerSig{newBaseBuiltinFunc(args, ctx)}
+	retType := c.inferType(args)
+	bf, err := newBaseBuiltinFuncWithTp(args, retType, ctx, tpString)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	sig := &builtinLowerSig{baseStringBuiltinFunc{bf}}
 	return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
 }
 
+func (c *lowerFunctionClass) inferType(args []Expression) *types.FieldType {
+	argTp := args[0].GetType()
```

Comment:
copied from concat.inferType ?
have you confirmed that the type inferring behavior 
of concat and upper are all the same?

If so, you can extract a function to avoid the duplicated code.